From ac85cae0941fa3904d2481031065845210e0a307 Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Sat, 12 Jul 2014 18:05:50 -0700
Subject: [PATCH 667/974] netdev: octeon-bgx-nexus: Only init port config data
 if probed.

Calling __cvmx_helper_init_port_config_data() if we don't have BGX/PKI
causes octeon-ethernet driver to fail.

Signed-off-by: David Daney <david.daney@cavium.com>
[Original patch taken from Cavium SDK 3.1.2-568]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 drivers/net/ethernet/octeon/octeon-bgx-nexus.c | 21 +++++++--------------
 1 file changed, 7 insertions(+), 14 deletions(-)

diff --git a/drivers/net/ethernet/octeon/octeon-bgx-nexus.c b/drivers/net/ethernet/octeon/octeon-bgx-nexus.c
index 24b42ce..0c16f38 100644
--- a/drivers/net/ethernet/octeon/octeon-bgx-nexus.c
+++ b/drivers/net/ethernet/octeon/octeon-bgx-nexus.c
@@ -38,6 +38,8 @@
 
 #include "octeon-bgx.h"
 
+static atomic_t bgx_nexus_once;
+
 static int bgx_probe(struct platform_device *pdev)
 {
 	struct bgx_platform_data platform_data;
@@ -53,6 +55,10 @@ static int bgx_probe(struct platform_device *pdev)
 	int r = 0;
 	char id[64];
 
+	/* One time initialization */
+	if (atomic_cmpxchg(&bgx_nexus_once, 0, 1) == 0)
+		__cvmx_helper_init_port_config_data();
+
 	reg = of_get_property(pdev->dev.of_node, "reg", NULL);
 	addr = of_translate_address(pdev->dev.of_node, reg);
 	interface = (addr >> 24) & 0xf;
@@ -138,20 +144,7 @@ void bgx_nexus_load(void)
 }
 EXPORT_SYMBOL(bgx_nexus_load);
 
-static int __init bgx_driver_init(void)
-{
-	int r;
-	__cvmx_helper_init_port_config_data();
-	r =  platform_driver_register(&bgx_driver);
-	return r;
-}
-module_init(bgx_driver_init);
-
-static void __exit bgx_driver_exit(void)
-{
-	platform_driver_unregister(&bgx_driver);
-}
-module_exit(bgx_driver_exit);
+module_platform_driver(bgx_driver);
 
 MODULE_LICENSE("GPL");
 MODULE_AUTHOR("Cavium Networks <support@caviumnetworks.com>");
-- 
2.6.2

